File: verifier/compact_witness.py
#!/usr/bin/env python3
"""Canonical parser for compact exact rational matrix witness files."""

from __future__ import annotations

import hashlib
import io
import math
import os
import stat
from dataclasses import dataclass
from fractions import Fraction
from pathlib import Path
from typing import BinaryIO, Callable, Iterable, Sequence


MAGIC = b"KNWIT003"


class CompactWitnessError(Exception):
    """Raised when compact witness bytes are malformed or noncanonical."""


@dataclass(frozen=True)
class WitnessLimits:
    max_bytes: int = 128 * 1024 * 1024
    max_blocks: int = 128
    max_dimension: int = 1024
    max_rank_sum: int = 10_000
    max_rationals: int = 1_000_000
    max_integer_bits: int = 131_072
    max_denominators_per_block: int = 1_000_000


DEFAULT_LIMITS = WitnessLimits()


@dataclass(frozen=True)
class BlockLayout:
    name: str
    dimension: int


@dataclass(frozen=True)
class LDLBlock:
    name: str
    dimension: int
    rank: int
    permutation: tuple[int, ...]
    diagonal: tuple[Fraction, ...]
    columns: tuple[tuple[Fraction, ...], ...]


@dataclass(frozen=True)
class MatrixBlock:
    name: str
    dimension: int
    matrix: tuple[tuple[Fraction, ...], ...]

    @property
    def rank(self) -> int:
        """Strict positivity verification certifies full rank."""

        return self.dimension


@dataclass(frozen=True)
class WitnessReport:
    sha256: str
    size_bytes: int
    block_count: int
    rank_sum: int
    rational_count: int
    maximum_numerator_bits: int
    maximum_denominator_bits: int


class _Reader:
    def __init__(self, stream: BinaryIO, limits: WitnessLimits) -> None:
        self.stream = stream
        self.limits = limits
        self.bytes_read = 0

    def read_exact(self, count: int, path: str) -> bytes:
        if count < 0:
            raise CompactWitnessError(f"{path}: negative read length")
        data = self.stream.read(count)
        self.bytes_read += len(data)
        if len(data) != count:
            raise CompactWitnessError(f"{path}: truncated witness")
        if self.bytes_read > self.limits.max_bytes:
            raise CompactWitnessError("witness exceeds byte limit")
        return data

    def read_u8(self, path: str) -> int:
        return self.read_exact(1, path)[0]

    def read_u16(self, path: str) -> int:
        return int.from_bytes(self.read_exact(2, path), "little")

    def read_u32(self, path: str) -> int:
        return int.from_bytes(self.read_exact(4, path), "little")

    def read_uleb128(self, path: str) -> int:
        value = 0
        shift = 0
        encoded = bytearray()
        while True:
            if len(encoded) >= 16:
                raise CompactWitnessError(f"{path}: ULEB128 is too long")
            byte = self.read_u8(path)
            encoded.append(byte)
            value |= (byte & 0x7F) << shift
            if byte < 0x80:
                break
            shift += 7
        canonical = bytearray()
        remaining = value
        while True:
            byte = remaining & 0x7F
            remaining >>= 7
            canonical.append(byte | (0x80 if remaining else 0))
            if not remaining:
                break
        if encoded != canonical:
            raise CompactWitnessError(f"{path}: noncanonical ULEB128")
        return value

    def read_magnitude(self, path: str) -> int:
        length = self.read_uleb128(f"{path}.length")
        if length == 0:
            raise CompactWitnessError(f"{path}: empty nonzero magnitude")
        max_bytes = (self.limits.max_integer_bits + 7) // 8
        if length > max_bytes:
            raise CompactWitnessError(f"{path}: integer exceeds bit limit")
        data = self.read_exact(length, path)
        if data[0] == 0:
            raise CompactWitnessError(f"{path}: magnitude has a leading zero")
        value = int.from_bytes(data, "big")
        if value.bit_length() > self.limits.max_integer_bits:
            raise CompactWitnessError(f"{path}: integer exceeds bit limit")
        return value

    def read_signed_integer(self, path: str) -> int:
        sign = self.read_u8(f"{path}.sign")
        if sign == 0:
            return 0
        if sign not in (1, 2):
            raise CompactWitnessError(f"{path}: invalid integer sign")
        value = self.read_magnitude(f"{path}.magnitude")
        return value if sign == 1 else -value


def _stable_regular_file(path: Path, limits: WitnessLimits) -> tuple[BinaryIO, os.stat_result]:
    flags = os.O_RDONLY
    flags |= getattr(os, "O_CLOEXEC", 0)
    flags |= getattr(os, "O_NOFOLLOW", 0)
    descriptor: int | None = None
    try:
        descriptor = os.open(path, flags)
        before = os.fstat(descriptor)
    except OSError as exc:
        if descriptor is not None:
            os.close(descriptor)
        raise CompactWitnessError(f"cannot open witness: {exc}") from exc
    if not stat.S_ISREG(before.st_mode):
        os.close(descriptor)
        raise CompactWitnessError("witness is not a regular file")
    if before.st_size > limits.max_bytes:
        os.close(descriptor)
        raise CompactWitnessError("witness exceeds byte limit")
    stream = os.fdopen(descriptor, "rb", closefd=True)
    return stream, before


def _expected_value_count(dimension: int) -> int:
    return dimension * (dimension + 1) // 2


def _read_block(
    reader: _Reader,
    layout: BlockLayout,
    block_index: int,
    limits: WitnessLimits,
    remaining_rank: int,
    remaining_rationals: int,
) -> tuple[MatrixBlock, int, int, int]:
    path = f"block[{block_index}]"
    dimension = reader.read_u16(f"{path}.dimension")
    if dimension != layout.dimension:
        raise CompactWitnessError(
            f"{path}: expected dimension {layout.dimension}, received {dimension}"
        )
    if dimension < 1 or dimension > limits.max_dimension:
        raise CompactWitnessError(f"{path}: invalid dimension")
    if dimension > remaining_rank:
        raise CompactWitnessError("rank-sum limit exceeded")
    value_count = _expected_value_count(dimension)
    if value_count > remaining_rationals:
        raise CompactWitnessError("rational-count limit exceeded")

    denominator_count = reader.read_u32(f"{path}.denominator_count")
    if denominator_count > limits.max_denominators_per_block:
        raise CompactWitnessError(f"{path}: too many denominators")
    if denominator_count == 0:
        raise CompactWitnessError(f"{path}: matrix block lacks denominators")
    if denominator_count > value_count:
        raise CompactWitnessError(f"{path}: denominator table contains unused values")

    denominators: list[int] = []
    previous = 0
    for index in range(denominator_count):
        value = reader.read_magnitude(f"{path}.denominators[{index}]")
        if value <= previous:
            raise CompactWitnessError(
                f"{path}: denominators must be unique and strictly increasing"
            )
        denominators.append(value)
        previous = value

    values: list[Fraction] = []
    used_denominators: set[int] = set()
    maximum_numerator_bits = 0
    maximum_denominator_bits = 0
    for index in range(value_count):
        denominator_index = reader.read_uleb128(
            f"{path}.values[{index}].denominator_index"
        )
        if denominator_index >= len(denominators):
            raise CompactWitnessError(
                f"{path}.values[{index}]: denominator index is out of range"
            )
        numerator = reader.read_signed_integer(
            f"{path}.values[{index}].numerator"
        )
        denominator = denominators[denominator_index]
        if math.gcd(abs(numerator), denominator) != 1:
            raise CompactWitnessError(
                f"{path}.values[{index}]: rational is not reduced"
            )
        values.append(Fraction(numerator, denominator))
        used_denominators.add(denominator_index)
        maximum_numerator_bits = max(
            maximum_numerator_bits, abs(numerator).bit_length()
        )
        maximum_denominator_bits = max(
            maximum_denominator_bits, denominator.bit_length()
        )
    if used_denominators != set(range(len(denominators))):
        raise CompactWitnessError(f"{path}: denominator table contains unused values")

    matrix = [
        [Fraction(0) for _ in range(dimension)]
        for _ in range(dimension)
    ]
    position = 0
    for row in range(dimension):
        for column in range(row, dimension):
            value = values[position]
            position += 1
            matrix[row][column] = value
            matrix[column][row] = value
    if position != len(values):
        raise CompactWitnessError(f"{path}: internal value-count mismatch")
    return (
        MatrixBlock(
            name=layout.name,
            dimension=dimension,
            matrix=tuple(tuple(row) for row in matrix),
        ),
        value_count,
        maximum_numerator_bits,
        maximum_denominator_bits,
    )


def _read_compact_witness_stream(
    stream: BinaryIO,
    size_bytes: int,
    expected_blocks: Sequence[BlockLayout],
    *,
    consumer: Callable[[MatrixBlock], None] | None = None,
    limits: WitnessLimits = DEFAULT_LIMITS,
) -> WitnessReport:
    if len(expected_blocks) > limits.max_blocks:
        raise CompactWitnessError("expected block layout exceeds block limit")
    if size_bytes > limits.max_bytes:
        raise CompactWitnessError("witness exceeds byte limit")
    digest = hashlib.sha256()

    class HashingStream:
        def read(self, count: int = -1) -> bytes:
            data = stream.read(count)
            digest.update(data)
            return data

    hashing_stream = HashingStream()
    reader = _Reader(hashing_stream, limits)
    rank_sum = 0
    rational_count = 0
    maximum_numerator_bits = 0
    maximum_denominator_bits = 0
    try:
        if reader.read_exact(len(MAGIC), "magic") != MAGIC:
            raise CompactWitnessError("incorrect witness magic")
        block_count = reader.read_u32("block_count")
        if block_count != len(expected_blocks):
            raise CompactWitnessError(
                f"expected {len(expected_blocks)} blocks, received {block_count}"
            )
        for block_index, layout in enumerate(expected_blocks):
            block, count, numerator_bits, denominator_bits = _read_block(
                reader,
                layout,
                block_index,
                limits,
                limits.max_rank_sum - rank_sum,
                limits.max_rationals - rational_count,
            )
            rank_sum += block.dimension
            rational_count += count
            maximum_numerator_bits = max(maximum_numerator_bits, numerator_bits)
            maximum_denominator_bits = max(
                maximum_denominator_bits, denominator_bits
            )
            if consumer is not None:
                consumer(block)
        if hashing_stream.read(1):
            raise CompactWitnessError("witness has trailing bytes")
    except OSError as exc:
        raise CompactWitnessError(f"cannot read witness: {exc}") from exc
    if reader.bytes_read != size_bytes:
        raise CompactWitnessError("witness size changed while reading")
    return WitnessReport(
        sha256=digest.hexdigest(),
        size_bytes=size_bytes,
        block_count=len(expected_blocks),
        rank_sum=rank_sum,
        rational_count=rational_count,
        maximum_numerator_bits=maximum_numerator_bits,
        maximum_denominator_bits=maximum_denominator_bits,
    )


def read_compact_witness_bytes(
    data: bytes,
    expected_blocks: Sequence[BlockLayout],
    *,
    consumer: Callable[[MatrixBlock], None] | None = None,
    limits: WitnessLimits = DEFAULT_LIMITS,
) -> WitnessReport:
    """Parse one immutable witness byte string."""

    if not isinstance(data, bytes):
        raise CompactWitnessError("witness data must be immutable bytes")
    return _read_compact_witness_stream(
        io.BytesIO(data),
        len(data),
        expected_blocks,
        consumer=consumer,
        limits=limits,
    )


def read_compact_witness(
    path: str | os.PathLike[str],
    expected_blocks: Sequence[BlockLayout],
    *,
    consumer: Callable[[MatrixBlock], None] | None = None,
    limits: WitnessLimits = DEFAULT_LIMITS,
) -> WitnessReport:
    """Parse a strict witness and optionally stream each block to a consumer."""

    witness_path = Path(path)
    stream, before = _stable_regular_file(witness_path, limits)
    try:
        report = _read_compact_witness_stream(
            stream,
            before.st_size,
            expected_blocks,
            consumer=consumer,
            limits=limits,
        )
        after = os.fstat(stream.fileno())
    finally:
        stream.close()
    if (
        before.st_dev,
        before.st_ino,
        before.st_size,
        before.st_mtime_ns,
    ) != (
        after.st_dev,
        after.st_ino,
        after.st_size,
        after.st_mtime_ns,
    ):
        raise CompactWitnessError("witness changed while reading")
    return report


def reconstruct_block(
    block: LDLBlock | MatrixBlock,
) -> tuple[tuple[Fraction, ...], ...]:
    """Reconstruct one original-coordinate matrix for tests and reference use."""

    if isinstance(block, MatrixBlock):
        if len(block.matrix) != block.dimension or any(
            len(row) != block.dimension for row in block.matrix
        ):
            raise CompactWitnessError(
                f"{block.name}: matrix dimensions are inconsistent"
            )
        if any(
            block.matrix[row][column] != block.matrix[column][row]
            for row in range(block.dimension)
            for column in range(row + 1, block.dimension)
        ):
            raise CompactWitnessError(f"{block.name}: matrix is not symmetric")
        return block.matrix

    permuted = [
        [Fraction(0) for _ in range(block.dimension)]
        for _ in range(block.dimension)
    ]
    for diagonal, column in zip(block.diagonal, block.columns):
        for row, left in enumerate(column):
            if not left:
                continue
            for other, right in enumerate(column):
                if right:
                    permuted[row][other] += diagonal * left * right
    result = [
        [Fraction(0) for _ in range(block.dimension)]
        for _ in range(block.dimension)
    ]
    for row, original_row in enumerate(block.permutation):
        for column, original_column in enumerate(block.permutation):
            result[original_row][original_column] = permuted[row][column]
    return tuple(tuple(row) for row in result)

File: verifier/cpp/compact_witness.cpp
#include "compact_witness.hpp"

#include <algorithm>
#include <array>
#include <cerrno>
#include <fcntl.h>
#include <limits>
#include <sstream>
#include <sys/stat.h>
#include <unistd.h>
#include <utility>

namespace knwit003 {
namespace {

constexpr std::array<std::uint8_t, 8> kMagic = {
    0x4b, 0x4e, 0x57, 0x49, 0x54, 0x30, 0x30, 0x33,
};

std::string indexed_path(
    const std::string& prefix,
    const std::size_t index) {
    return prefix + "[" + std::to_string(index) + "]";
}

std::size_t bit_length(const mpz_class& value) {
    if (value == 0) {
        return 0;
    }
    return mpz_sizeinbase(value.get_mpz_t(), 2);
}

class FileDescriptor final {
  public:
    explicit FileDescriptor(const int descriptor) : descriptor_(descriptor) {}
    FileDescriptor(const FileDescriptor&) = delete;
    FileDescriptor& operator=(const FileDescriptor&) = delete;
    ~FileDescriptor() {
        if (descriptor_ >= 0) {
            static_cast<void>(::close(descriptor_));
        }
    }

    int get() const {
        return descriptor_;
    }

  private:
    int descriptor_;
};

bool same_file_identity(const struct stat& left, const struct stat& right) {
    if (left.st_dev != right.st_dev ||
        left.st_ino != right.st_ino ||
        left.st_size != right.st_size) {
        return false;
    }
#if defined(__APPLE__)
    return left.st_mtimespec.tv_sec == right.st_mtimespec.tv_sec &&
           left.st_mtimespec.tv_nsec == right.st_mtimespec.tv_nsec &&
           left.st_ctimespec.tv_sec == right.st_ctimespec.tv_sec &&
           left.st_ctimespec.tv_nsec == right.st_ctimespec.tv_nsec;
#else
    return left.st_mtim.tv_sec == right.st_mtim.tv_sec &&
           left.st_mtim.tv_nsec == right.st_mtim.tv_nsec &&
           left.st_ctim.tv_sec == right.st_ctim.tv_sec &&
           left.st_ctim.tv_nsec == right.st_ctim.tv_nsec;
#endif
}

class Reader final {
  public:
    explicit Reader(const std::span<const std::uint8_t> bytes) : bytes_(bytes) {
        if (bytes_.size() > kMaximumBytes) {
            throw ParseError("witness exceeds byte limit");
        }
    }

    std::span<const std::uint8_t> read_exact(
        const std::size_t count,
        const std::string& path) {
        if (count > bytes_.size() - position_) {
            throw ParseError(path + ": truncated witness");
        }
        const auto result = bytes_.subspan(position_, count);
        position_ += count;
        return result;
    }

    std::uint8_t read_u8(const std::string& path) {
        return read_exact(1, path)[0];
    }

    std::uint16_t read_u16(const std::string& path) {
        const auto data = read_exact(2, path);
        return static_cast<std::uint16_t>(
            static_cast<std::uint16_t>(data[0]) |
            static_cast<std::uint16_t>(
                static_cast<std::uint16_t>(data[1]) << 8U));
    }

    std::uint32_t read_u32(const std::string& path) {
        const auto data = read_exact(4, path);
        return static_cast<std::uint32_t>(
            static_cast<std::uint32_t>(data[0]) |
            (static_cast<std::uint32_t>(data[1]) << 8U) |
            (static_cast<std::uint32_t>(data[2]) << 16U) |
            (static_cast<std::uint32_t>(data[3]) << 24U));
    }

    mpz_class read_uleb128(const std::string& path) {
        mpz_class value = 0;
        unsigned long shift = 0;
        for (std::size_t index = 0; index < kMaximumUleb128Bytes; ++index) {
            const std::uint8_t byte = read_u8(path);
            const std::uint8_t payload =
                static_cast<std::uint8_t>(byte & 0x7fU);
            value += mpz_class(payload) << shift;
            if ((byte & 0x80U) == 0U) {
                if (index > 0U && payload == 0U) {
                    throw ParseError(path + ": noncanonical ULEB128");
                }
                return value;
            }
            shift += 7U;
        }
        throw ParseError(path + ": ULEB128 is too long");
    }

    std::size_t read_bounded_uleb128(
        const std::string& path,
        const std::size_t maximum) {
        const mpz_class value = read_uleb128(path);
        if (value > maximum) {
            throw ParseError(path + ": value exceeds limit");
        }
        return value.get_ui();
    }

    mpz_class read_magnitude(const std::string& path) {
        constexpr std::size_t maximum_bytes =
            (kMaximumIntegerBits + 7U) / 8U;
        const std::size_t length =
            read_bounded_uleb128(path + ".length", maximum_bytes);
        if (length == 0U) {
            throw ParseError(path + ": empty nonzero magnitude");
        }
        const auto data = read_exact(length, path);
        if (data.front() == 0U) {
            throw ParseError(path + ": magnitude has a leading zero");
        }

        mpz_class value = 0;
        mpz_import(
            value.get_mpz_t(),
            data.size(),
            1,
            1,
            1,
            0,
            data.data());
        if (bit_length(value) > kMaximumIntegerBits) {
            throw ParseError(path + ": integer exceeds bit limit");
        }
        return value;
    }

    mpz_class read_signed_integer(const std::string& path) {
        const std::uint8_t sign = read_u8(path + ".sign");
        if (sign == 0U) {
            return 0;
        }
        if (sign != 1U && sign != 2U) {
            throw ParseError(path + ": invalid integer sign");
        }
        mpz_class value = read_magnitude(path + ".magnitude");
        if (sign == 2U) {
            value = -value;
        }
        return value;
    }

    bool at_end() const {
        return position_ == bytes_.size();
    }

  private:
    std::span<const std::uint8_t> bytes_;
    std::size_t position_ = 0;
};

std::size_t expected_value_count(const std::size_t dimension) {
    return dimension * (dimension + 1U) / 2U;
}

struct ParsedBlock {
    MatrixBlock block;
    std::size_t value_count;
    std::size_t maximum_numerator_bits;
    std::size_t maximum_denominator_bits;
};

ParsedBlock read_block(
    Reader& reader,
    const BlockLayout& layout,
    const std::size_t block_index,
    const std::size_t remaining_rank_budget,
    const std::size_t remaining_rational_budget) {
    const std::string path = indexed_path("block", block_index);
    const std::size_t dimension = reader.read_u16(path + ".dimension");

    if (dimension != layout.dimension) {
        std::ostringstream message;
        message << path << ": expected dimension " << layout.dimension
                << ", received " << dimension;
        throw ParseError(message.str());
    }
    if (dimension == 0U || dimension > kMaximumDimension) {
        throw ParseError(path + ": invalid dimension");
    }
    if (dimension > remaining_rank_budget) {
        throw ParseError("rank-sum limit exceeded");
    }

    const std::size_t value_count = expected_value_count(dimension);
    if (value_count > remaining_rational_budget) {
        throw ParseError("rational-count limit exceeded");
    }

    const std::size_t denominator_count =
        reader.read_u32(path + ".denominator_count");
    if (denominator_count == 0U) {
        throw ParseError(path + ": matrix block lacks denominators");
    }
    if (denominator_count > kMaximumDenominatorsPerBlock) {
        throw ParseError(path + ": too many denominators");
    }
    if (denominator_count > value_count) {
        throw ParseError(path + ": denominator table contains unused values");
    }

    std::vector<mpz_class> denominators;
    denominators.reserve(denominator_count);
    mpz_class previous = 0;
    for (std::size_t index = 0; index < denominator_count; ++index) {
        const mpz_class value =
            reader.read_magnitude(indexed_path(path + ".denominators", index));
        if (value <= previous) {
            throw ParseError(
                path +
                ": denominators must be unique and strictly increasing");
        }
        denominators.push_back(value);
        previous = value;
    }

    std::vector<mpq_class> values;
    values.reserve(value_count);
    std::vector<bool> used_denominators(denominator_count, false);
    std::size_t maximum_numerator_bits = 0;
    std::size_t maximum_denominator_bits = 0;
    for (std::size_t index = 0; index < value_count; ++index) {
        const std::string value_path = indexed_path(path + ".values", index);
        const std::size_t denominator_index = reader.read_bounded_uleb128(
            value_path + ".denominator_index",
            denominator_count);
        if (denominator_index >= denominator_count) {
            throw ParseError(
                value_path + ": denominator index is out of range");
        }
        const mpz_class numerator =
            reader.read_signed_integer(value_path + ".numerator");
        const mpz_class& denominator = denominators[denominator_index];

        const mpz_class numerator_magnitude =
            numerator < 0 ? -numerator : numerator;
        mpz_class divisor = 0;
        mpz_gcd(
            divisor.get_mpz_t(),
            numerator_magnitude.get_mpz_t(),
            denominator.get_mpz_t());
        if (divisor != 1) {
            throw ParseError(value_path + ": rational is not reduced");
        }

        mpq_class value(numerator, denominator);
        value.canonicalize();
        values.push_back(std::move(value));
        used_denominators[denominator_index] = true;
        maximum_numerator_bits =
            std::max(maximum_numerator_bits, bit_length(numerator_magnitude));
        maximum_denominator_bits =
            std::max(maximum_denominator_bits, bit_length(denominator));
    }
    if (std::find(
            used_denominators.begin(),
            used_denominators.end(),
            false) != used_denominators.end()) {
        throw ParseError(path + ": denominator table contains unused values");
    }

    Matrix matrix(
        dimension,
        std::vector<mpq_class>(dimension, mpq_class(0)));
    std::size_t position = 0;
    for (std::size_t row = 0; row < dimension; ++row) {
        for (std::size_t column = row; column < dimension; ++column) {
            matrix[row][column] = values[position];
            matrix[column][row] = values[position];
            ++position;
        }
    }
    if (position != values.size()) {
        throw ParseError(path + ": internal value-count mismatch");
    }

    return ParsedBlock{
        MatrixBlock{
            layout.name,
            dimension,
            std::move(matrix),
        },
        value_count,
        maximum_numerator_bits,
        maximum_denominator_bits,
    };
}

}  // namespace

WitnessReport parse_compact_witness(
    const std::span<const std::uint8_t> bytes,
    const std::span<const BlockLayout> expected_blocks,
    const BlockConsumer& consumer) {
    if (expected_blocks.size() > kMaximumBlocks) {
        throw ParseError("expected block layout exceeds block limit");
    }

    Reader reader(bytes);
    const auto magic = reader.read_exact(kMagic.size(), "magic");
    if (!std::equal(magic.begin(), magic.end(), kMagic.begin())) {
        throw ParseError("incorrect witness magic");
    }

    const std::size_t block_count = reader.read_u32("block_count");
    if (block_count > kMaximumBlocks) {
        throw ParseError("block-count limit exceeded");
    }
    if (block_count != expected_blocks.size()) {
        std::ostringstream message;
        message << "expected " << expected_blocks.size()
                << " blocks, received " << block_count;
        throw ParseError(message.str());
    }

    std::size_t rank_sum = 0;
    std::size_t rational_count = 0;
    std::size_t maximum_numerator_bits = 0;
    std::size_t maximum_denominator_bits = 0;
    for (std::size_t block_index = 0;
         block_index < expected_blocks.size();
         ++block_index) {
        ParsedBlock parsed = read_block(
            reader,
            expected_blocks[block_index],
            block_index,
            kMaximumRankSum - rank_sum,
            kMaximumRationals - rational_count);
        rank_sum += parsed.block.dimension;
        rational_count += parsed.value_count;
        maximum_numerator_bits =
            std::max(maximum_numerator_bits, parsed.maximum_numerator_bits);
        maximum_denominator_bits =
            std::max(maximum_denominator_bits, parsed.maximum_denominator_bits);
        if (consumer) {
            consumer(std::move(parsed.block));
        }
    }

    if (!reader.at_end()) {
        throw ParseError("witness has trailing bytes");
    }
    return WitnessReport{
        bytes.size(),
        block_count,
        rank_sum,
        rational_count,
        maximum_numerator_bits,
        maximum_denominator_bits,
    };
}

WitnessReport parse_compact_witness_file(
    const std::filesystem::path& path,
    const std::span<const BlockLayout> expected_blocks,
    const BlockConsumer& consumer) {
    int flags = O_RDONLY;
#if defined(O_CLOEXEC)
    flags |= O_CLOEXEC;
#endif
#if defined(O_NOFOLLOW)
    flags |= O_NOFOLLOW;
#endif
    const int raw_descriptor = ::open(path.c_str(), flags);
    if (raw_descriptor < 0) {
        throw ParseError("cannot open witness");
    }
    const FileDescriptor descriptor(raw_descriptor);

    struct stat before {};
    if (::fstat(descriptor.get(), &before) != 0) {
        throw ParseError("cannot determine witness metadata");
    }
    if (!S_ISREG(before.st_mode)) {
        throw ParseError("witness is not a regular file");
    }
    if (before.st_size < 0) {
        throw ParseError("witness size is invalid");
    }
    const auto file_size = static_cast<std::uintmax_t>(before.st_size);
    if (file_size > kMaximumBytes) {
        throw ParseError("witness exceeds byte limit");
    }
    if (file_size > std::numeric_limits<std::size_t>::max()) {
        throw ParseError("witness size is not addressable");
    }

    std::vector<std::uint8_t> bytes(static_cast<std::size_t>(file_size));
    std::size_t position = 0;
    while (position < bytes.size()) {
        const ssize_t count = ::read(
            descriptor.get(),
            bytes.data() + position,
            bytes.size() - position);
        if (count < 0 && errno == EINTR) {
            continue;
        }
        if (count <= 0) {
            throw ParseError("witness changed while reading");
        }
        position += static_cast<std::size_t>(count);
    }
    std::uint8_t trailing = 0;
    ssize_t trailing_count = 0;
    do {
        trailing_count = ::read(descriptor.get(), &trailing, 1);
    } while (trailing_count < 0 && errno == EINTR);
    if (trailing_count != 0) {
        throw ParseError("witness changed while reading");
    }

    struct stat after {};
    if (::fstat(descriptor.get(), &after) != 0 ||
        !same_file_identity(before, after)) {
        throw ParseError("witness changed while reading");
    }
    return parse_compact_witness(bytes, expected_blocks, consumer);
}

Matrix reconstruct_block(const MatrixBlock& block) {
    if (block.dimension == 0U || block.dimension > kMaximumDimension) {
        throw ParseError("block has invalid reconstruction dimension");
    }
    if (block.matrix.size() != block.dimension) {
        throw ParseError("block has inconsistent matrix dimensions");
    }
    for (std::size_t row = 0; row < block.dimension; ++row) {
        if (block.matrix[row].size() != block.dimension) {
            throw ParseError("block has inconsistent matrix dimensions");
        }
        for (std::size_t column = row + 1U;
             column < block.dimension;
             ++column) {
            if (block.matrix[row][column] != block.matrix[column][row]) {
                throw ParseError("block matrix is not symmetric");
            }
        }
    }
    return block.matrix;
}

}  // namespace knwit003
